### PR TITLE
2018 migration

### LINKED ISSUE #1180: Finish 2018 migration
After 2019-01-31 we should finish our migration to the 2018 edition.  This will make Amethyst no longer compatible with Rust compilers < 1.31, which is why I propose we wait until 1.32 is stable.  That should be happening on 2019-01-31.

### PR DESCRIPTION
**Pull request by Xaeroxe**
Welcome to Rust 2018!  Closes #1180

### CODE REVIEW

File: examples/animation/main.rs [hunk 1]
```diff
@@ -47,16 +47,16 @@ impl Default for Example {
...
         // Initialise the scene with an object, a light and a camera.
-        let prefab_handle = world.exec(|loader: PrefabLoader<MyPrefabData>| {
+        let prefab_handle = world.exec(|loader: PrefabLoader<'_, MyPrefabData>| {
             loader.load("prefab/animation.ron", RonFormat, (), ())
         });
         self.sphere = Some(world.create_entity().with(prefab_handle).build());
     }
 
-    fn handle_event(&mut self, data: StateData<GameData>, event: StateEvent) -> SimpleTrans {
+    fn handle_event(&mut self, data: StateData<'_, GameData<'_, '_>>, event: StateEvent) -> SimpleTrans {
```
Comment (jojolepro) [inline examples/animation/main.rs]: wtf

Why did they remove implicit lifetimes :(

**Comment by Xaeroxe** [comment]
You uh might want to review this one [here](https://github.com/amethyst/amethyst/pull/1224/files/aa72dc2bb665298fdce5cfb343d0405c4991b23c) due to the litany of formatting changes 😁

**Comment by Xaeroxe** [comment]
Oh heh I guess `cargo fix` also applied some unrelated fixes for `amethyst_gltf` since it had a few warnings.  That's a neat subcommand!

File: examples/animation/main.rs (hunk 1 above)
Comment (Xaeroxe) [reply examples/animation/main.rs]: They're still implicit cuz `'_`.  They're just explicitly implicit 😜

**Comment by Xaeroxe** [comment]
We can get rid of the `#[macro_use]` by `use`ing the macros we need, but I figured for this iteration I should focus on getting us compiling on 1.31 with no warnings as quick and easy as possible.

**Comment by azriel91** [comment]
There's a number of diffs like this:

```patch
-extern crate amethyst;
+use amethyst;

use amethyst::{
    ...
};
```

Do people think those are in scope of this PR?
Those `use`s are likely superfluous.

**Comment by azriel91** [comment]
Worth mentioning in `README.md` and `docs/CHANGELOG.md` that the minimum Rust version is now 1.31.0

**Comment by azriel91** [comment]
bors r=jojolepro,LucioFranco,termhn,azriel91

**Comment by jojolepro** [comment]
Missing book changes. The book still contains the old way of creating states (implicit lifetimes).

**Comment by LucioFranco** [comment]
Going to merge this as I have a mostly working branch with the book updates and I will open a PR soon.

bors r=jojolepro,LucioFranco,termhn,azriel91

**Comment by LucioFranco** [comment]
book PR is open #1226.

**Pull request merged**